### PR TITLE
Unified module loader support in binaryen.js

### PR DESCRIPTION
**Pull request by dcodeIO**
This PR adds support for AMD loaders to the binaryen.js wrapper and unifies existing exposure and loader code. It accounts for the checks performed in `check.py` (generated `a.js` is assumed to provide `Binaryen` as a global even in a CommonJS environment) by simply exporting globally in any case. In practice, this shouldn't lead to any issues and shortens the code quite a bit.

Additionally, new but not yet referenced source files that currently generate warnings in compilation have been added to `build-js.sh` and `bin/binaryen.js` / `bin/wasm.js` have been recompiled based on the changes.

The underlying reason for this change is that I am trying to build a single UMD-compatible distribution for one of my projects and it's currently necessary to provide [special configuration](https://github.com/dcodeIO/AssemblyScript/blob/master/dist/check-amd.html#L6) for just binaryen.js. A [similar PR](https://github.com/WebAssembly/wabt/pull/536) has been made to WABT (especially libwabt.js) as well, bringing both in line.

### CODE REVIEW

File: src/js/binaryen.js-post.js [hunk 1]
```diff
@@ -911,34 +911,12 @@
     return Module['_BinaryenSetAPITracing'](on);
   };
 
-  return Module;
-};
+  // Always expose globally because check.py generates a.js which requires 'Binaryen' as a global var
```
Comment (kripken) [inline src/js/binaryen.js-post.js]: if this is just for check.py, perhaps we could do some hack in there, just for testing? seems a shame to instrument the official production build for testing.

File: src/js/binaryen.js-post.js [hunk 2]
```diff
@@ -911,34 +911,12 @@
...
+  (typeof global != "undefined" && global || typeof window != "undefined" && window || this)["Binaryen"] = Module;
```
Comment (kripken) [inline src/js/binaryen.js-post.js]: this statement could use some parentheses, to clarify the order of operations.

File: src/js/binaryen.js-pre.js [hunk 3]
```diff
@@ -1 +1,2 @@
-var Binaryen = function(Module) {
```
Comment (kripken) [inline src/js/binaryen.js-pre.js]: doesn't this define Binaryen as a global symbol?

Comment (dcodeIO) [inline src/js/binaryen.js-pre.js]: Changed this initially just to put everything into a closure. Now that a.js is instrumented instead of the production build and the Binaryen global exposed conditionally, it's actually necessary.

File: check.py [hunk 4]
```diff
@@ -413,6 +413,7 @@ def fix(x):
     print s
     f = open('a.js', 'w')
     f.write(open(os.path.join(options.binaryen_bin, 'binaryen.js')).read())
+    f.write('var Binaryen = module.exports;\n')
```
Comment (kripken) [inline check.py]: Sorry to go back around on this, but this troubles me. I'd like binaryen.js to "just work" in the easiest way for people. That means

1. If you include it on your web page or bundle it alongside your JS code, you can just call `Binaryen.*`, without needing that line of boilerplate.
2. If you use it in a unified module loader context, it works in a consistent way with that approach.

So I would like to avoid adding this in `check.py`, and having it in binaryen.js seems natural for 1. Is it impossible to achieve both 1 and 2 at the same time?

Comment (dcodeIO) [inline check.py]: The way it works now is:

1. [Check if binaryen.js is run in a CommonJS context](https://github.com/dcodeIO/binaryen/blob/4f62b7fd88a4f59338af2b26dd46cf55c6a909f7/src/js/binaryen.js-post.js#L914). If so, export `Module`
2. [Check if binaryen.js is run in an AMD context](https://github.com/dcodeIO/binaryen/blob/4f62b7fd88a4f59338af2b26dd46cf55c6a909f7/src/js/binaryen.js-post.js#L916). If so, define a factory method that returns `Module`
3. [Otherwise, expose Binaryen as a global var](https://github.com/dcodeIO/binaryen/blob/4f62b7fd88a4f59338af2b26dd46cf55c6a909f7/src/js/binaryen.js-post.js#L918)

The reason why the additional line is necessary in check.py is that the generated a.js is running inside of a CommonJS context, but not used like that because the tests are just concatenated. That's merely a special case of the test suite.

Comment (kripken) [inline check.py]: I see. Could we fix `check.py` to use binaryen in the normal way for a CommonJS context? It would be good if the test suite reflected real-world usage.

Comment (dcodeIO) [inline check.py]: Can try. This depends on whether there is a `require` function with file system access in 'MozJS', respectively what else is the intended way to load CommonJS modules there. If that's possible, `a.js` could just `var Binaryen = require("./bin/binaryen.js");` the distribution instead of copying its contents.

Comment (dcodeIO) [inline check.py]: As I am unable to run the tests locally because I am on Windows and the Travis build is also failing lately, things become a bit difficult here.

I assume that 'mozjs' is just a recent Spidermonkey shell, is that correct? If so, the version I am running locally doesn't seem to provide a CommonJS environment like the initial comment in `binaryen.js-post.js`, that lead me to adding the additional line to `check.py`, stated (there's no `module` or `require` defined). If the latter assumption is correct, the additional line added to `check.py` wouldn't be necessary, but hard to tell without a way to test this.

**Comment by dcodeIO** [comment]
As I am unable to run the tests locally because I am on Windows and the Travis build is also failing lately, things become a bit difficult here.

I assume that 'mozjs' is just a recent Spidermonkey shell, is that correct? If so, the version I am running locally doesn't seem to provide a CommonJS environment like the initial comment in `binaryen.js-post.js`, that lead me to adding the additional line to `check.py`, stated (there's no `module` or `require` defined). If the latter assumption is correct, the additional line added to `check.py` wouldn't be necessary, but hard to tell without a way to test this.

File: check.py (hunk 4 above)
Comment (kripken) [inline check.py]: Hmm, good point, mozjs does not have CommonJS functionality. We could use node.js instead, but it would need a very recent node, we'd need to version-detect, etc. So all that seems out of scope for this PR I guess.

File: src/js/binaryen.js-post.js [hunk 5]
```diff
@@ -911,34 +911,11 @@
     return Module['_BinaryenSetAPITracing'](on);
   };
 
-  return Module;
-};
+  if (typeof module !== "undefined" && module && module.exports)
```
Comment (kripken) [inline src/js/binaryen.js-post.js]: some comments explaining the things done here would help I think

**Comment by kripken** [comment]
I am concerned that before we had `var Binaryen =` which meant that bundling the code inside other JS was possible, and it just added one local var (which may have been non-global). With this PR, we would add a global var. Could we assign just the local var, like before, for the non-module loader mode?

also cc @yurydelendik who knows about this stuff

**Comment by dcodeIO** [comment]
Exposing globally alongside exporting as a module doesn't, in my experience, cause any practical problems usually and I am also fine with this approach for my specific use case.

When there is a `var Binaryen = (...)()`, though, it'll always declare this variable, no matter the context (might be local, might be global) and no matter the return value of the closure (even when it returns nothing, there'll be a `Binaryen` var evaluating to `undefined`).

Hence, where avoiding a global is of importance for some reason, people came up with the (nowadays relatively standard) fallback in question that checks for `global`, `window` and `this` in this order. `global` is node-like and also catches other *custom* environments reliably,, `window` browser-like and `this` usually catches everything else like running within a web worker or the `a.js` use case we have here.

No matter how you decide, I'll happily change it to fit your needs.

**Comment by kripken** [comment]
Ok, good points. Let's do it your way.
